doc/decisions/warning-array.md

```diff
@@ -0,0 +1,36 @@
+# Arrays
+
+## Problem
+
+Currently multiple warnings are saved in an elektra non-conforming array
+notation which is limited to 100 entries. The notation of `#00` is against
+the design [decision made](array.md).
+
+Furthermore if the number of warnings is exceeded, the earliest entries are overwritten
```

Comment:
This is not a problem?

Reply:
We discussed this with @kodebach in an elektra meeting that this is a problem.

Reply:
The problem are situations like this:
```
#01: the newest warning
#02: the oldest warning
...
#99:  the second newest warning
```

So that new/old warnings might get mixed up.

There are several ways how to avoid it, one way is to simply continue counting up and start dropping the warnings with the lowest index.

I remember that @kodebach said that we should also add a warning when we started doing something like this.

Reply:
How can this happen?
What do you mean by "dropping the warnings with the lowest index"?
Should it be possible to have only entries from `#3` to `#99` for example because 0-2 is dropped?

Reply:
Exactly. (`#99` would be `#_99`, though)

Reply:
As far as I remember we agreed on this:

Handle the first `n-1` warnings as normal. Instead of the `n`-th warning we issue a special warning that the limit has been reached and ignore the actual warning. From then on all warnings are dropped. Additionally we could simply log all warnings above the limit, so that in case of debugging the warnings are not completely gone.

The reason for this behaviour is that often the first warning created is the most important to fix the problem. Later warnings may be caused by earlier ones, if the earlier warning is overwritten it becomes hard to diagnose.

Reply:
> Exactly. (#99 would be #_99, though)

Why exactly the underscore? This is really confusing.
I am also confused why we should only partially drop entries. What I have seen in the code is that you iterate over an elektra array and also save the current array number under `warnings/number`.

> Instead of the n-th warning

What exactly is `n` here? @markus2330 does not want a hard cap but this obviously requires one.

Reply:
> Additionally we could simply log all warnings above the limit, so that in case of debugging the warnings are not completely gone.

In debugging you usually only have short sessions (not many kdbGet in a long-running loop) thus it is rather unlikely you will reach the limit.

> The reason for this behaviour is that often the first warning created is the most important to fix the problem.

This is true for warnings within a single execution of `kdb*`-call (e.g. `kdbSet`) In applications many warnings might get collected until finally an error occurs and everything is presented to the user. Then the *most recent* warnings are relevant. Not what happened a long time ago.

Alternatively, we could simply clean all warnings before a `kdb*`-call. Then the behavior as described by @kodebach is best (only keep the earliest warnings). This decision does not mention this, however.

> Later warnings may be caused by earlier ones, if the earlier warning is overwritten it becomes hard to diagnose.

Yes, within a single kdb*-call it is like this.

> Why exactly the underscore?

It is needed for correct sorting.

> This is really confusing.

You are free to implement #2698. It would be highly appreciated.

Reply:
I am much more confused now than before. I still don't know why you simply drop complete warnings when the current warning number is saved in the parent meta (`warning/number`). When and how to use these weird underscores? Is there any document which explains this? I will stick to the array number of 100 because there is neither better a reason to use 1000 nor to use 10.
What exactly is the desired behavior now when `n` is exceeded?
I feel like consuming more time from you rather if someone else implements this with more insight in Elektra ...

Reply:
> When and how to use these weird underscores?

Simply put: Elektra sorts lexicographically using ASCII codes. `_` has a higher ASCII code than the digits `0`-`9`. Therefore by prefixing the index with `_` we ensure that the sort order is `#0`, `#1`, `#2`, ..., `#9`, `#_10` instead of `#0`, `#1`, `#10`, `#2`, ..., `#9`. You need more underscores, the longer the number is because otherwise the same problem occurs for 10 and 100 instead of 1 and 10.

> Is there any document which explains this?

Yes, https://github.com/ElektraInitiative/libelektra/blob/master/doc/tutorials/arrays.md. Feel free to improve it, if it isn't clear enough.

> What exactly is the desired behavior now when n is exceeded?

For now, I think we should stick to the old behaviour of overwriting old warnings (so basically the index is always modulo `n`), because of the problem with multiple kdb*-calls.